### ISSUE TITLE
Regression: Return of the double-flash bug

### ISSUE THREAD

**Issue by nerdinand** [hanami/controller#260]
It seems #196 has returned to haunt us.

Here's a minimal test app to reproduce this:
[flash_test.zip](https://github.com/hanami/controller/files/1937644/flash_test.zip)

1. Visit http://localhost:2300/. You will see:
![20180423_121750](https://user-images.githubusercontent.com/113440/39120949-6423dc84-46f0-11e8-8311-9adde3ebe13b.png)
2. Click on Edit. You will see a form:
![20180423_121850](https://user-images.githubusercontent.com/113440/39120989-8027b6f8-46f0-11e8-809a-4b7001885b71.png)
3. Click on Save. As expected, the form will render again and show the success flash message:
![20180423_121943](https://user-images.githubusercontent.com/113440/39121024-9e7d799e-46f0-11e8-91d0-58317c26b186.png)
4. Click on Cancel. The root page will render but show the flash message again!
![20180423_122043](https://user-images.githubusercontent.com/113440/39121070-c2c31aca-46f0-11e8-80e9-d158e25bdbf8.png)

Expected behaviour: The root page renders without showing the flash message.

As with #196, adding
```ruby
before { flash.clear }
```
to the index action is a workaround.

**Comment by GustavoCaso** [comment]
I'm having a look at this one.

Looks like the current behavior store the flash from the last request I imagine is to access it for redirects. 

Maybe we can have a way of marking when a value has been accessed, and since flash message are meant for one use we could delete then in the next iteration.

WDYT? @jodosha

**Comment by GustavoCaso** [comment]
I have looked into a little more looks like the behavior is different depending on a redirect or a new request.

By adding this to the `Update` action from the sample app:
```ruby
def call(params)
  flash[:success_notice] = "Saved successfully."
  self.status = 201
  redirect_to '/'
end
```

Works as expected.

@jodosha I'm guessing the users need to experience same behavior is they are using `redirect` or normal request cycle, right?

**Comment by GustavoCaso** [comment]
@nerdinand we have confirmed that problem exists. Thank you for opening the issue ❤️ 

We are working on the solution. I will try opening a PR ASAP.

**Comment by jodosha** [comment]
Fixed by #267
